### PR TITLE
Animated Menu

### PR DESCRIPTION
**Pull request by Torpedo99**
# Description
This PR adds an Animated Menu to the game.

# Testing
Open the game and watch the Passerby ship form and move across the ship, followed by the Hostile Ship which shoots asteroids. (Hostile ship is continuously repeated)

# Outstanding Work
- Add Shoot mechanism
- Dynamic stars

# Notes
- The current PR has issues with asteroid size. Sizes of 0.4 are drawn way too big for their size.
- Collision issues since the body is a bit bigger and malformed. This may be due to the fact that asteroids are very small (0.2), which is not supported by the MeshLoader.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
This pull request **introduces 2 alerts** when merging d206d866b424781653ffe16eca28bf8d4962d928 into f5b74588fbe234229493abed7c59101ab4d492af - [view on LGTM.com](https://lgtm.com/projects/g/MovingBlocks/DestinationSol/rev/pr-bb637e45e238185c4bb5925bad3b34e463935671)

**new alerts:**

* 2 for Container contents are never accessed

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: engine/src/main/java/org/destinationsol/menu/BackgroundAsteroid.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public Vector2 getPosition() {
+        return position;
+    }
+
+    public float getAngle() {
+        return angle;
+    }
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/BackgroundAsteroid.java]: Functions should probably have Javadoc comments (specifically the non-getter functions). Also an unnecessary space here

File: engine/src/main/java/org/destinationsol/menu/BackgroundAsteroidManager.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        for (BackgroundAsteroid BackgroundAsteroid : BackgroundAsteroids) {
+            BackgroundAsteroid.update();
+
+            float distance = (float)Math.sqrt(Math.pow(BackgroundAsteroid.getPosition().x - displayDimensions.getRatio()/2, 2) + Math.pow(BackgroundAsteroid.getPosition().y - 0.5f, 2));
+            if (distance < radius) {
+                retainedBackgroundAsteroids.add(BackgroundAsteroid);
+            } else {
+                retainedBackgroundAsteroids.add(buildAsteroid());
+            }
+        }
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/BackgroundAsteroidManager.java]: Include a comment (or two) for this block of code.

File: engine/src/main/java/org/destinationsol/menu/BackgroundAsteroidManager.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+        Body body = asteroidMeshLoader.getBodyAndSprite(texture, scale*0.905f, BodyDef.BodyType.DynamicBody, position, angle, new ArrayList<>(), 10f, DrawableLevel.BODIES);
+        body.setLinearVelocity(velocity);
+        body.setAngularVelocity(angularVelocity);
+        BackgroundAsteroid asteroid = new BackgroundAsteroid(texture, scale, tint, position, velocity, angle, body);
+        body.setUserData(asteroid);
+
+        return asteroid;
+    }
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/BackgroundAsteroidManager.java]: You could probably decompose this function into smaller functions. makes it a bit easier to understand that way. Also, unnecessary space here.

File: engine/src/main/java/org/destinationsol/menu/BackgroundShip.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    private float angle;
+
+    private float orgX;
+    private float orgY;
+
+    private Color tint;
+
+    private Body body;
+
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/BackgroundShip.java]: Unnecessary space.

File: engine/src/main/java/org/destinationsol/menu/BackgroundShip.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    public void draw(UiDrawer drawer) {
+        drawer.draw(texture, scale, scale, orgX, orgY, position.x, position.y, angle, tint);
+    }
+
+    public Vector2 getPosition() {
+        return position;
+    }
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/BackgroundShip.java]: This class looks very similar to `BackgroundAsteroid`, maybe you could have them inherit from a new class `BackgroundObject`?

File: engine/src/main/java/org/destinationsol/menu/MainMenuScreen.java [hunk 6]
```diff
@@ -116,6 +136,9 @@ public void updateCustom(SolApplication solApplication, SolInputManager.InputPoi
         if (creditsControl.isJustOff()) {
             inputManager.setScreen(solApplication, screens.credits);
         }
+
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MainMenuScreen.java]: unnecessary space

File: engine/src/main/java/org/destinationsol/menu/MainMenuScreen.java [hunk 7]
```diff
@@ -144,4 +169,5 @@ public void drawImages(UiDrawer uiDrawer, SolApplication solApplication) {
     private int calculateButtonOffsetFromBottom(int buttonIndex) {
         return -(buttonPadding + buttonHeight / 2) - (buttonIndex * (buttonPadding + buttonHeight));
     }
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MainMenuScreen.java]: Not needed.

File: engine/src/main/java/org/destinationsol/menu/MenuContactListener.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    @Override
+    public void preSolve(Contact contact, Manifold oldManifold) {
+    }
+
+    @Override
+    public void postSolve(Contact contact, ContactImpulse impulse) {
+
+    }
+
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MenuContactListener.java]: Unnecessary space.

File: engine/src/main/java/org/destinationsol/menu/MenuContactListener.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+package org.destinationsol.menu;
+
+import com.badlogic.gdx.physics.box2d.Contact;
+import com.badlogic.gdx.physics.box2d.ContactImpulse;
+import com.badlogic.gdx.physics.box2d.ContactListener;
+import com.badlogic.gdx.physics.box2d.Manifold;
+
+public class MenuContactListener implements ContactListener {
+
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MenuContactListener.java]: Unnecessary space.

File: engine/src/main/java/org/destinationsol/menu/MenuMeshLoader.java [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        JSONObject rootNode = json.getJsonValue();
+
+        Validator.validate(rootNode, "engine:schemaCollisionMesh");
+
+        readModel(rootNode);
+
+        json.dispose();
+    }
+
+
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MenuMeshLoader.java]: Unnecessary space.

File: engine/src/main/java/org/destinationsol/menu/MenuMeshLoader.java [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+    public static class PolygonModel {
+        public final List<Vector2> vertices = new ArrayList<>();
+        private Vector2[] tmpArray; // Used to avoid allocation in attachFixture()
+    }
+
+    public static class CircleModel {
+        public final Vector2 center = new Vector2();
+        public float radius;
+    }
+}
```
Comment (bpas247) [inline engine/src/main/java/org/destinationsol/menu/MenuMeshLoader.java]: Javadoc comments needed for at least the most complex functions in this class.

**Comment by NicholasBatesNZ** [comment]
What's the status on this @Torpedo99 ? Do you plan to continue work or should we close this?

**Comment by Torpedo99** [comment]
@NicholasBatesNZ I do plan to work on this further, just that I'm having trouble getting enough time. (School takes up 12 hours of the day). I'll be skipping school a few days so I'll try to get this finished before the GCI starts.

**Comment by NicholasBatesNZ** [comment]
Closing due to more work being unlikely, as discussed on chat

**Pull request closed without merging**